src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+package util;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class DeferredProperty<T> {
+    /**
+     * The temporary stored value of the property
+     */
+    private T value;
+
+    /**
+     * If true then the {@link DeferredProperty#value} needs to be updated.
+     */
+    private boolean isDirty;
+
+    /**
+     * The supplier used to autoamtically compute {@link DeferredProperty#value} when {@link DeferredProperty#isDirty} is true.
+     */
+    private final Supplier<T> supplier;
+
+    /**
+     * The collection of {@link DeferredProperty} instances that should be made dirty when this instance gets dirty.
+     */
+    private final List<DeferredProperty<?>> observers;
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     * If no initial value is provided (i.e. <code>null</code> value) then this instance will start of as dirty.
+     * Otherwise, if a none <code>null</code> value is provided then it will start of clean.
```

Comment:
```suggestion
     * If no initial value is provided (i.e. {@code null} value) then this instance will start of as dirty.
     * Otherwise, if a none {@code null} value is provided then it will start of clean.
```

Maybe you can make the phrasing slightly easier to follow:
```suggestion
     * If the provided initial value is {@code null}, then this instance will start of as dirty.
     * Otherwise, it will start clean with the provided initial value.
```